File: android-frontend/app/src/main/java/com/example/schedulingapp_android/viewmodels/TaskViewModel.kt
package com.example.schedulingapp_android.viewmodels

import android.util.Log
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.schedulingapp_android.data.api.RetrofitClient
import com.example.schedulingapp_android.data.models.Task
import com.example.schedulingapp_android.data.models.TaskCreateRequest
import com.example.schedulingapp_android.data.models.TaskCreateResponse
import com.example.schedulingapp_android.data.models.TaskUpdateRequest
import kotlinx.coroutines.launch

class TaskViewModel : ViewModel() {
    private val _tasks = MutableLiveData<List<Task>>()
    val tasks: LiveData<List<Task>> get() = _tasks

    fun getAllTasksForUser(userId: Int) {
        viewModelScope.launch {
            try {
                val response = RetrofitClient.apiService.getAllTasks(userId)
                if (response.isSuccessful) {
                    _tasks.postValue(response.body())
                } else {
                    _tasks.postValue(emptyList())
                    Log.e("TaskViewModel", "Failed to fetch user's tasks: ${response.errorBody()?.string()}")
                }
            } catch (e: Exception) {
                _tasks.postValue(emptyList())
                Log.e("TaskViewModel", "Error fetching user's tasks: ${e.message}", e)
            }
        }
    }

    fun getNotStartedTasks(userId: Int) {
        viewModelScope.launch {
            try {
                val response = RetrofitClient.apiService.getTasksNotStarted(userId)
                if (response.isSuccessful) {
                    _tasks.postValue(response.body())
                } else {
                    _tasks.postValue(emptyList())
                }
            } catch (e: Exception) {
                _tasks.postValue(emptyList())
            }
        }
    }

    fun getInProgressTasks(userId: Int) {
        viewModelScope.launch {
            try {
                val response = RetrofitClient.apiService.getTasksInProgress(userId)
                if (response.isSuccessful) {
                    _tasks.postValue(response.body())
                } else {
                    _tasks.postValue(emptyList())
                }
            } catch (e: Exception) {
                _tasks.postValue(emptyList())
            }
        }
    }

    fun getCompletedTasks(userId: Int) {
        viewModelScope.launch {
            try {
                val response = RetrofitClient.apiService.getCompletedTasks(userId)
                if (response.isSuccessful) {
                    _tasks.postValue(response.body())
                } else {
                    _tasks.postValue(emptyList())
                }
            } catch (e: Exception) {
                _tasks.postValue(emptyList())
            }
        }
    }

    fun createTask(userId: Int, task: TaskCreateRequest) {
        viewModelScope.launch {
            try {
                val response = RetrofitClient.apiService.createTask(userId, task)
                if (response.isSuccessful) {
                    val taskCreateResponse = response.body()
                    taskCreateResponse?.let {
                        val newTaskId = it.taskId

                        val taskResponse = RetrofitClient.apiService.getTaskById(userId, newTaskId)
                        if (taskResponse.isSuccessful) {
                            val newTask = taskResponse.body()

                            newTask?.let {
                                val updatedTaskList =
                                    _tasks.value?.toMutableList() ?: mutableListOf()
                                updatedTaskList.add(it)
                                _tasks.postValue(updatedTaskList)
                            }
                        } else {
                            Log.e("TaskViewModel", "Failed to fetch newly created task: ${taskResponse.errorBody()?.string()}")
                        }
                    } ?: run {
                        Log.e("TaskViewModel", "Failed to get taskId from response")
                    }
                } else {
                    Log.e("TaskViewModel", "Failed to create task: ${response.errorBody()?.string()}")
                }
            } catch (e: Exception) {
                Log.e("TaskViewModel", "Error creating task: ${e.message}", e)
            }
        }
    }

    fun editTask(userId: Int, taskId: Int, taskUpdateRequest: TaskUpdateRequest) {
        viewModelScope.launch {
            try {
                val response = RetrofitClient.apiService.updateTask(userId, taskId, taskUpdateRequest)
                if (response.isSuccessful) {
                    getAllTasksForUser(userId)
                } else {
                    Log.e("TaskViewModel", "Failed to edit task: ${response.errorBody()?.string()}")
                }
            } catch (e: Exception) {
                Log.e("TaskViewModel", "Error editing task: ${e.message}", e)
            }
        }
    }

    fun deleteTask(userId: Int, taskId: Int) {
        viewModelScope.launch {
            try {
                val response = RetrofitClient.apiService.deleteTask(userId, taskId)
                if (response.isSuccessful) {
                    _tasks.postValue(_tasks.value?.filter { it.id != taskId })
                } else {
                    Log.e("TaskViewModel", "Failed to delete task: ${response.errorBody()?.string()}")
                }
            } catch (e: Exception) {
                Log.e("TaskViewModel", "Error deleting task: ${e.message}", e)
            }
        }
    }
}

File: Scheduling/Scheduling/ViewModels/TaskViewModel.swift
//
//  TaskViewModel.swift
//  Scheduling
//
//  Created by Glace on 2025-04-07.
//

import Foundation

class TaskViewModel {
    private(set) var tasks: [Task] = []
    var onTasksUpdated: (() -> Void)?
    var onError: ((String) -> Void)?
    
    func getAllTasksForUser(userId: Int) {
        DataService.shared.getAllTasks(userId: userId) { [weak self] result in
            switch result {
            case .success(let tasks):
                self?.tasks = tasks
                self?.onTasksUpdated?()
            case .failure(_):
                self?.onError?("Failed to fetch tasks")
            }
        }
    }
    
    func getTasksByStatus(userId: Int, status: String) {
        DataService.shared.getTasksByStatus(userId: userId, status: status) { [weak self] result in
            switch result {
            case .success(let tasks):
                self?.tasks = tasks
                self?.onTasksUpdated?()
            case .failure(_):
                self?.onError?("Failed to fetch tasks")
            }
        }
    }
    
    func createTask(userId: Int, request: TaskCreateRequest, completion: @escaping (Bool) -> Void) {
        DataService.shared.createTask(userId: userId, request: request) { [weak self] result in
            switch result {
            case .success(_):
                self?.getAllTasksForUser(userId: userId)
                completion(true)
            case .failure(_):
                self?.onError?("Failed to create task")
                completion(false)
            }
        }
    }
    
    func updateTask(userId: Int, request: TaskUpdateRequest, completion: @escaping (Bool) -> Void) {
        DataService.shared.updateTask(userId: userId, request: request) { [weak self] result in
            switch result {
            case .success(_):
                self?.getAllTasksForUser(userId: userId)
                completion(true)
            case .failure(_):
                self?.onError?("Failed to update task")
                completion(false)
            }
        }
    }
    
    func deleteTask(userId: Int, taskId: Int) {
        DataService.shared.deleteTask(userId: userId, taskId: taskId) { [weak self] result in
            switch result {
            case .success(_):
                self?.tasks.removeAll { $0.id == taskId }
                self?.onTasksUpdated?()
            case .failure(_):
                self?.onError?("Failed to delete task")
            }
        }
    }
}
